server/manager/src/main/java/org/apache/accumulo/manager/upgrade/UpgradeCoordinator.java

```diff
@@ -142,11 +145,17 @@ public synchronized void upgradeZookeeper(ServerContext context,
         "Not currently in a suitable state to do zookeeper upgrade %s", status);
 
     try {
-      int cv = context.getServerDirs()
-          .getAccumuloPersistentVersion(context.getVolumeManager().getFirst());
-      ServerContext.ensureDataVersionCompatible(cv);
+      int cv = AccumuloDataVersion.getCurrentVersion(context);
       this.currentVersion = cv;
 
+      int oldestVersion = upgraders.entrySet().iterator().next().getKey();
+      if (cv < oldestVersion) {
+        String oldRelease = dataVersionToReleaseName(oldestVersion);
+        throw new UnsupportedOperationException("Upgrading from a version less than " + oldRelease
+            + " data version (" + oldestVersion + ") is not supported. Upgrade to at least "
+            + oldRelease + " before upgrading to " + Constants.VERSION);
+      }
+
```

Comment:
In thinking about this some more, I don't think this added message is ever going to work... this can only execute after the `UpgradeCoordinator` is instantiated... which won't happen until after the `Manager` is constructed... but the `Manager` won't successfully get constructed because `ServerContext.init` will fail long before this, when it runs `ServerContext.ensureDataVersionCompatible`, that checks `AccumuloDataVersion.CAN_RUN`.

I think if you want to have this more informative message, you're going to need to edit `ServerContext.ensureDataVersionCompatible` instead and this added check can just be removed, as it's not doing anything useful here.

Reply:
I'll need to check - but pretty sure that during development I had a bad version set and the errors in the logs prompted me to put this in.  Maybe because I and the new data version implements I was able to get this far, but I did reach the check.

Reply:
You are correct.  Going from 1.10 to 3.0-SNAPSHOT logs

```
2023-02-07T21:11:03,556 [start.Main] ERROR: Thread 'manager' died.
java.lang.IllegalStateException: This version of accumulo (3.0.0-SNAPSHOT) is not compatible with files stored using data version 8
        at org.apache.accumulo.server.ServerContext.ensureDataVersionCompatible(ServerContext.java:302) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.ServerContext.init(ServerContext.java:371) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.AbstractServer.<init>(AbstractServer.java:50) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.<init>(Manager.java:412) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.main(Manager.java:406) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.ManagerExecutable.execute(ManagerExecutable.java:45) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.start.Main.lambda$execKeyword$0(Main.java:122) ~[accumulo-start-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]

```

Reply:
@ctubbsii would this be okay? Seems long and not adding much - might just be better to include something in the release notes or in the upgrading section of the docs.

```
java.lang.IllegalStateException: This version of accumulo (3.0.0-SNAPSHOT) is not compatible with files stored using data version 8 see org.apache.accumulo.server.AccumuloDataVersion javadoc for historic version mapping
        at org.apache.accumulo.server.ServerContext.ensureDataVersionCompatible(ServerContext.java:302) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.ServerContext.init(ServerContext.java:372) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.AbstractServer.<init>(AbstractServer.java:50) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.<init>(Manager.java:412) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.main(Manager.java:406) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.ManagerExecutable.execute(ManagerExecutable.java:45) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.start.Main.lambda$execKeyword$0(Main.java:122) ~[accumulo-start-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]

```

Reply:
I think we can get what you were going for by updating that current error message to include what you were trying to include here. I can take a look and add that change to this PR if you want.

Reply:
Sure - a quick look and the info I wanted is either private or in the javadoc.

Reply:
a5da59abc76 removes the test and leaves the message alone.  

The improvement that I was shooting for would provide guidance in the manage log when the upgrade fails.  The upgrade is using data version and while that is tied to release version, that connection is buried in our code.  Users may not be familiar with "data version", where it is set / coming from and how that even relates to a release.

Reply:
Here's my proposed changes:

```diff
diff --git a/server/base/src/main/java/org/apache/accumulo/server/AccumuloDataVersion.java b/server/base/src/main/java/org/apache/accumulo/server/AccumuloDataVersion.java
index 9f523e7082..5b44913f11 100644
--- a/server/base/src/main/java/org/apache/accumulo/server/AccumuloDataVersion.java
+++ b/server/base/src/main/java/org/apache/accumulo/server/AccumuloDataVersion.java
@@ -86,7 +86,11 @@ public class AccumuloDataVersion {
     return cv;
   }
 
-  public static String dataVersionToReleaseName(final int version) {
+  public static String oldestUpgradeableVersionName() {
+    return dataVersionToReleaseName(CAN_RUN.stream().mapToInt(x -> x).min().orElseThrow());
+  }
+
+  private static String dataVersionToReleaseName(final int version) {
     switch (version) {
       case ROOT_TABLET_META_CHANGES:
         return "2.1.0";
diff --git a/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java b/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
index 03ab556923..e2286115c2 100644
--- a/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
+++ b/server/base/src/main/java/org/apache/accumulo/server/ServerContext.java
@@ -300,7 +300,9 @@ public class ServerContext extends ClientContext {
   public static void ensureDataVersionCompatible(int dataVersion) {
     if (!AccumuloDataVersion.CAN_RUN.contains(dataVersion)) {
       throw new IllegalStateException("This version of accumulo (" + Constants.VERSION
-          + ") is not compatible with files stored using data version " + dataVersion);
+          + ") is not compatible with files stored using data version " + dataVersion
+          + ". Please upgrade from " + AccumuloDataVersion.oldestUpgradeableVersionName()
+          + " or later.");
     }
   }
 
```

Reply:
Added in 2aea7e1f8d9 - will run an upgrade sequence and include the output for completeness later.

Reply:
With the changes - the error message now reads:

```2023-02-08T14:15:02,313 [start.Main] ERROR: Thread 'manager' died.
java.lang.IllegalStateException: This version of accumulo (3.0.0-SNAPSHOT) is not compatible with files stored using data version 8. Please upgrade from 2.1.0 or later.
        at org.apache.accumulo.server.ServerContext.ensureDataVersionCompatible(ServerContext.java:304) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.ServerContext.init(ServerContext.java:373) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.server.AbstractServer.<init>(AbstractServer.java:50) ~[accumulo-server-base-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.<init>(Manager.java:412) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.Manager.main(Manager.java:406) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.manager.ManagerExecutable.execute(ManagerExecutable.java:45) ~[accumulo-manager-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at org.apache.accumulo.start.Main.lambda$execKeyword$0(Main.java:122) ~[accumulo-start-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]


```